plugins/audio-s5gen2/fu-audio-s5gen2-ble-device.c

```diff
@@ -0,0 +1,615 @@
+
+static gboolean
+fu_qc_s5gen2_ble_device_notify_acquire_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(device);
+
+	if (!fu_qc_s5gen2_ble_device_notify_release(self, error))
+		return FALSE;
+
+	return fu_qc_s5gen2_ble_device_notify_acquire(self, error);
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_probe(FuDevice *device, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(device);
+	g_autofree gchar *vendor_id = NULL;
+
+	if (!FU_DEVICE_CLASS(fu_qc_s5gen2_ble_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* after reboot the device might appear too fast */
+	if (!fu_device_retry_full(device,
+				  fu_qc_s5gen2_ble_device_notify_acquire_cb,
+				  FU_QC_S5GEN2_BLE_DEVICE_AQUIRE_RETRIES,
+				  FU_QC_S5GEN2_BLE_DEVICE_TIMEOUT,
+				  NULL,
+				  error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_get_api(self, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_get_features(self, FALSE, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_get_serial(self, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_get_variant(self, error))
+		return FALSE;
+
+	if (self->feature.core >= 2)
+		if (!fu_qc_s5gen2_ble_device_set_transport_protocol(
+			self,
+			FU_QC_S5GEN2_GAIA_PROTOCOL_VERSION,
+			error))
+			return FALSE;
+
+	/* set vendor ID to avoid update error */
+	vendor_id = g_strdup_printf("BLUETOOTH:%04X", self->vid_v3);
+	fu_device_add_vendor_id(device, vendor_id);
+
+	if (!fu_qc_s5gen2_ble_device_register_notification(self, error))
+		return FALSE;
+
+	return fu_qc_s5gen2_ble_device_notify_release(FU_QC_S5GEN2_BLE_DEVICE(device), error);
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_set_quirk_kv(FuDevice *device,
+				     const gchar *key,
+				     const gchar *value,
+				     GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(device);
+	guint64 tmp = 0;
+
+	if (g_strcmp0(key, "AudioS5gen2Gaia3VendorId") == 0) {
+		if (!fu_strtoull(value, &tmp, 0, G_MAXUINT16, FU_INTEGER_BASE_AUTO, error))
+			return FALSE;
+		self->vid_v3 = tmp;
+		return TRUE;
+	}
+
+	/* failed */
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "quirk key not supported");
+	return FALSE;
+}
+
+static void
+fu_qc_s5gen2_ble_device_init(FuQcS5gen2BleDevice *self)
+{
+	self->vid_v3 = FU_QC_S5GEN2_GAIA_V3_DEFAULT_VENDOR;
+	self->mtu = 0;
+	self->feature.core = 0;
+	self->feature.dfu = 0;
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_PLAIN);
+	fu_device_set_remove_delay(FU_DEVICE(self), FU_QC_S5GEN2_DEVICE_REMOVE_DELAY);
+	fu_device_add_protocol(FU_DEVICE(self), "com.qualcomm.s5gen2");
+}
+
+static void
+fu_qc_s5gen2_ble_device_finalize(GObject *object)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(object);
+	g_autoptr(GError) error = NULL;
+
+	if (self->io_cmd != NULL) {
+		g_object_unref(self->io_cmd);
+		self->io_cmd = NULL;
+		self->mtu = 0;
```

Comment:
i don't think you need do do this in `finalize`, only if you're using `destroy`.

Reply:
Right, just a good practice, even if unneded sometimes. removed